### PR TITLE
Bluetooth: controller: Add ll_adv_aux_set_clear function

### PR DESCRIPTION
**Pull request by dagbja**
Add ll_adv_aux_set_clear, add to bsim test

Signed-off-by: Dag Bjarvin <Dag.Bjarvin@nordicsemi.no>

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/ull_adv_aux.c [hunk 1]
```diff
@@ -51,6 +51,12 @@ static struct ll_adv_aux_set ll_adv_aux_pool[CONFIG_BT_CTLR_ADV_AUX_SET];
 static void *adv_aux_free;
 #endif /* (CONFIG_BT_CTLR_ADV_AUX_SET > 0) */
 
+#if defined(CONFIG_BT_CTLR_ADV_SET)
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:54]: If this define is duplicated in ull_adv.c file. Have this in ull_adv_internal.h file.

File: subsys/bluetooth/controller/ll_sw/ull_adv_aux.c [hunk 2]
```diff
@@ -595,10 +601,49 @@ uint8_t ll_adv_aux_set_remove(uint8_t handle)
 
 uint8_t ll_adv_aux_set_clear(void)
 {
-	/* TODO: reset/release all adv set primary channel and  Aux channel
-	 * PDUs
-	 */
-	return 0;
+	struct ll_adv_set *adv;
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:604]: Any reasons why `ll_adv_aux_set_remove`  cant be called in a loop?

**Review by cvinayak — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_adv_aux.c (hunk 1 above)
Comment (dagbja) [reply subsys/bluetooth/controller/ll_sw/ull_adv_aux.c]: Fixed.

**Comment by dagbja** [comment]
I think I accidentally pushed a wrong button and lost your second comment.
In ll_adv_aux_set_remove, if adv set is not created, we return with an error.
In ll_adv_aux_set_clear, we continue to next set.
We could create a new function with the core functionality, and a flag to decide action for sets not created, and call this function for both remove and clear. But I don't think it is worth the performance penalty.

File: subsys/bluetooth/controller/ll_sw/ull_adv_aux.c [hunk 3]
```diff
@@ -595,10 +595,49 @@ uint8_t ll_adv_aux_set_remove(uint8_t handle)
...
+
+	for (handle = 0; handle < BT_CTLR_ADV_SET; ++handle) {
+		/* Get the advertising set instance */
+		adv = ull_adv_is_created_get(handle);
+
+		if (!adv) {
+			continue;
+		}
+
+		if (adv->is_enabled) {
```
Comment (kruithofa) [inline subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:610]: I know the spec is unclear about the behaviour here, but what happens now is that we release advertiser sets up to the first one that is enabled, and do not release the remaining ones.
So for example if we have 3 instances (1, 2 and 3) and nr. 2 is still enabled we remove nr. 1, and return error-code BT_HCI_ERR_CMD_DISALLOWED, but instance nr. 3 remains untouched.
Wouldn't it be more consistent to either remove all instances (that are not enabled), or to keep all instances?

**Review by kruithofa — CHANGES_REQUESTED** [review]

**Comment by kruithofa** [comment]
> I think I accidentally pushed a wrong button and lost your second comment.
> In ll_adv_aux_set_remove, if adv set is not created, we return with an error.
> In ll_adv_aux_set_clear, we continue to next set.
> We could create a new function with the core functionality, and a flag to decide action for sets not created, and call this function for both remove and clear. But I don't think it is worth the performance penalty.

Can't you do something like:
**for (handle=0; handle<BT_CTLR_ADV_SET; handle++) {
err = ll_adv_aux_set_remove(handle);
if (err == BT_HCI_ERR_CMD_DISALLOWED)
    return err;
}**

File: tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c [hunk 4]
```diff
@@ -545,6 +545,38 @@ static void test_advx_main(void)
...
+	printk("success.\n");
+
+	printk("Trying to remove adv sets ...");
+	for (handle = 0; handle < num_adv_sets; ++handle) {
+		err = ll_adv_aux_set_remove(handle);
+		if (err != BT_HCI_ERR_UNKNOWN_ADV_IDENTIFIER) {
+			goto exit;
+		}
+	}
+	printk("success.\n");
```
Comment (kruithofa) [inline tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:578]: We are missing a test for the case that you try to clear all sets while one or more are still enabled

**Review by kruithofa — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_adv_aux.c (hunk 3 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:610]: We should remove all, or none. Which implies, 2 step process, of check for all sets being disabled, and then remove all. ` ll_adv_aux_set_remove` be refactored into 2 smaller functions and reused here in 2 loops (check loop, and remove loop).

**Comment by cvinayak** [comment]
> I think I accidentally pushed a wrong button and lost your second comment.

You did "resolve conversation", I have again "unresolve... " it now :-)

File: subsys/bluetooth/controller/ll_sw/ull_adv_aux.c (hunk 2 above)
Comment (dagbja) [reply subsys/bluetooth/controller/ll_sw/ull_adv_aux.c]: Changed to call ll_adv_aux_set_remove in a loop.

File: subsys/bluetooth/controller/ll_sw/ull_adv_aux.c (hunk 3 above)
Comment (dagbja) [reply subsys/bluetooth/controller/ll_sw/ull_adv_aux.c]: Changed to disable all enabled sets, return error if one or more was enabled.

File: tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c (hunk 4 above)
Comment (dagbja) [reply tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:578]: Changed test to enable every other set, that should cover the case of only some enabled.

Comment (kruithofa) [reply tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:578]: Still missing the case where the function returns BT_HCI_ERR_CMD_DISALLOWED

Comment (dagbja) [reply tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:578]: Add a test for that.

Comment (dagbja) [reply tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:578]: Added a test for that.

**Review by wopu-ot — APPROVED** [review]

**Review by kruithofa — APPROVED** [review]

**Comment by carlescufi** [comment]
@cvinayak please take another look

File: subsys/bluetooth/controller/ll_sw/ull_adv_aux.c [hunk 5]
```diff
@@ -595,10 +595,18 @@ uint8_t ll_adv_aux_set_remove(uint8_t handle)
 
 uint8_t ll_adv_aux_set_clear(void)
 {
-	/* TODO: reset/release all adv set primary channel and  Aux channel
-	 * PDUs
-	 */
-	return 0;
+	uint8_t handle;
+	uint8_t err;
+	uint8_t retval = BT_HCI_ERR_SUCCESS;
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:600]: Reverse Christmas tree

**Review by cvinayak — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_adv_aux.c (hunk 5 above)
Comment (dagbja) [reply subsys/bluetooth/controller/ll_sw/ull_adv_aux.c]: Fixed

**Comment by carlescufi** [comment]
@dagbja please clean up the commits, you have a merge commit in there.
@cvinayak please re-review

**Review by carlescufi — APPROVED** [review]

**Review by cvinayak — APPROVED** [review]

**Pull request merged**